### PR TITLE
[coding-standards] Add fixers to autocomplete missing return and param annotations

### LINKED ISSUE #417: Annotations for all methods [US-4596]
To improve code uniformity and simplify orientation in the code, it would be nice to ensure the consistency of the annotations.

### What is happening
Some methods or attributes are annotated and some are not.

### Expected result
- Every method that is not without any parameters and without return value is annotated
- The annotation contains all parameters and their types (eg. `@param int $productId`) and the return value (`@return void` is not used).
- The package shopsys/coding-standards should include a automatic checker (or, preferably, a fixer) that enforce this rule.
- The change is documented and the contribution guides are reviewed.

*Note: this is a duplicate of the US-4596 in our internal backlog.*

### PR DESCRIPTION
**Pull request by TomasVotruba**
| Q             | A
| ------------- | ---
|Description, reason for the PR| #417, Yodiz: [US-4596](https://app.yodiz.com/plan/pages/detail.vz?cid=10273#/us-4596)
|New feature| Yes
|BC breaks| No
|Fixes issues| closes #417 
|Standards and tests pass| Yes
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

This PR contains **the fixers only**.

The **coding standard fix** PR ([preview](https://github.com/shopsys/shopsys/commit/d4e43c9bdc531d757ab8953f269770ecdba3e9d0)) will follow in **standalone PR to keep codereview clear here**.

### 1. How to handle No-type Cases?  :heavy_check_mark: 

```diff
+    /**
+     * @param mixed $name
+     */
     function someFunction($name)
```

### 2. How to handle arrays?

Current:

```diff
+    /**
+     * @param array $names
+     */
     function someFunction($names)
```

`array` or `mixed[]`? 

Having `mixed[]` is works better with Slevomat sniffs/PHPStan. The `@param array` is often reported as **weak type**.

```diff
+    /**
+     * @param mixed[] $names
+     */
     function someFunction($names)
```

### 3. FQN `@var`, `@param` and `@return` corrections  :heavy_check_mark: 

Solved by Slevomat sniff:
   
```yaml
services:
    SlevomatCodingStandard\Sniffs\Namespaces\FullyQualifiedClassNameInAnnotationSniff: ~
```

There are some non-fqn cases already + PHPStorm does this:
  
gifit

### CODE REVIEW

File: packages/coding-standards/src/Helper/PhpToDocTypeTransformer.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+namespace Shopsys\CodingStandards\Helper;
+
+use Nette\Utils\Strings;
+use PhpCsFixer\Tokenizer\Analyzer\Analysis\TypeAnalysis;
+use PhpCsFixer\Tokenizer\Tokens;
+
+final class PhpToDocTypeTransformer
+{
+    /**
+     * @var FqnNameResolver
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply packages/coding-standards/src/Helper/PhpToDocTypeTransformer.php]: You can type such comments just once to save us time, it's clear.

I'll add sniff to solve this.

File: packages/coding-standards/src/Helper/PhpToDocTypeTransformer.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+     */
+    private $fqnNameResolver;
+
+    public function __construct()
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply packages/coding-standards/src/Helper/PhpToDocTypeTransformer.php]: I don't get it

File: packages/coding-standards/src/Helper/PhpdocRegex.php [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+declare(strict_types=1);
+
+namespace Shopsys\CodingStandards\Helper;
+
+final class PhpdocRegex
+{
+    /**
+     * "@param typ $value"
+     * ↓
+     * "$value"
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply packages/coding-standards/src/Helper/PhpdocRegex.php]: It's example how this method works. Input -> output.

I'll make it more clear

File: packages/coding-standards/src/Helper/ShopsysFixerNaming.php [hunk 4]
```diff
@@ -0,0 +1,29 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Shopsys\CodingStandards\Helper;
+
+use Nette\Utils\Strings;
+
+final class ShopsysFixerNaming
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply packages/coding-standards/src/Helper/ShopsysFixerNaming.php]: I will add some description.

File: packages/coding-standards/src/Helper/FqnNameResolver.php [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+                return $namespaceUseAnalysis->getFullName();
+            }
+        }
+
+        $namespaceTokens = $tokens->findGivenKind([T_NAMESPACE], 0);
+        if (!count($namespaceTokens[T_NAMESPACE])) {
+            return $className;
+        }
+
+        $namespaceToken = array_pop($namespaceTokens);
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply packages/coding-standards/src/Helper/FqnNameResolver.php]: I'll wrap it, it's doing to much

File: packages/coding-standards/src/Helper/FqnNameResolver.php [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+        $namespaceUseAnalyses = $this->namespaceUsesAnalyzer->getDeclarationsFromTokens($tokens);
+        foreach ($namespaceUseAnalyses as $namespaceUseAnalysis) {
+            if ($className === $namespaceUseAnalysis->getShortName()) {
+                return $namespaceUseAnalysis->getFullName();
+            }
+        }
+
+        $namespaceTokens = $tokens->findGivenKind([T_NAMESPACE], 0);
+        if (!count($namespaceTokens[T_NAMESPACE])) {
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply packages/coding-standards/src/Helper/FqnNameResolver.php]: Why is that better?

File: packages/coding-standards/src/CsFixer/Phpdoc/OrderedParamAnnotationsFixer.php [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+        $matches = Strings::match($line->getContent(), PhpdocRegex::ARGUMENT_NAME_PATTERN);
+        if ($matches[1]) {
+            return $matches[1];
+        }
+
+        return null;
+    }
+
+    /**
+     * @param int[] $argumentAnalyses
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply packages/coding-standards/src/CsFixer/Phpdoc/OrderedParamAnnotationsFixer.php]: Yes,will fix

File: packages/coding-standards/src/CsFixer/Phpdoc/OrderedParamAnnotationsFixer.php [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+            return $lines;
+        }
+
+        $argumentNamesToKeys = $this->resolveArgumentNamesToKeys($argumentAnalyses, $docParamNamesToKeys);
+        $properParamOrder = array_merge($docParamNamesToKeys, $argumentNamesToKeys);
+
+        $newLines = [];
+        foreach ($lines as $position => $line) {
+            $paramName = $this->getParamNameFromLine($line);
+            if ($paramName && isset($properParamOrder[$paramName])) {
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply packages/coding-standards/src/CsFixer/Phpdoc/OrderedParamAnnotationsFixer.php]: No troubles.

It's longer version with same meaning in this context.
Why is it better?

File: packages/coding-standards/src/CsFixer/Phpdoc/OrderedParamAnnotationsFixer.php [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+        return ShopsysFixerNaming::createFromClass(self::class);
+    }
+
+    /**
+     * Needs to run after @see MissingParamAnnotationsFixer
+     * @return int
+     */
+    public function getPriority(): int
+    {
+        return -5;
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply packages/coding-standards/src/CsFixer/Phpdoc/OrderedParamAnnotationsFixer.php]: It's good to keep space for other fixers.

File: packages/coding-standards/src/CsFixer/Phpdoc/OrderedParamAnnotationsFixer.php [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+            if ($this->shouldSkipDocToken($token)) {
+                continue;
+            }
+
+            $functionTokenPosition = $tokens->getNextTokenOfKind($index, [new Token([T_FUNCTION, 'function'])]);
+            if (!$functionTokenPosition) {
+                continue;
+            }
+
+            $argumentAnalyses = $this->functionsAnalyzer->getFunctionArguments($tokens, $functionTokenPosition);
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply packages/coding-standards/src/CsFixer/Phpdoc/OrderedParamAnnotationsFixer.php]: Why? That seems to vague and could be anything.

The name here respects returned object type.

File: packages/coding-standards/src/CsFixer/Phpdoc/MissingParamAnnotationsFixer.php [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    /**
+     * @return \PhpCsFixer\FixerDefinition\FixerDefinitionInterface
+     */
+    public function getDefinition(): FixerDefinitionInterface
+    {
+        return new FixerDefinition(
+            'Methods and functions has to have @param annotation for all params and @return annotation',
+            [new CodeSample(
+                <<<'SAMPLE'
+function someFunction(int $value)
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply packages/coding-standards/src/CsFixer/Phpdoc/MissingParamAnnotationsFixer.php]: How would you know what the fixer does without "after" example?

These are mostly useless if not part ofofficial php-cs-fixer, but it's the best way to explain programmers what it does. Without looking into tests

File: packages/coding-standards/src/CsFixer/Phpdoc/MissingParamAnnotationsFixer.php [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+final class MissingParamAnnotationsFixer extends AbstractMissingAnnotationsFixer
+{
+    /**
+     * @return \PhpCsFixer\FixerDefinition\FixerDefinitionInterface
+     */
+    public function getDefinition(): FixerDefinitionInterface
+    {
+        return new FixerDefinition(
+            'Methods and functions has to have @param annotation for all params and @return annotation',
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply packages/coding-standards/src/CsFixer/Phpdoc/MissingParamAnnotationsFixer.php]: I'll fix it. Forgotten

File: packages/coding-standards/src/CsFixer/Phpdoc/AbstractMissingAnnotationsFixer.php [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+    {
+        return str_repeat($this->whitespacesFixerConfig->getIndent(), $this->indentDetector->detectOnPosition($tokens, $index));
+    }
+
+    /**
+     * @param \PhpCsFixer\Tokenizer\Tokens $tokens
+     * @param int $index
+     * @return bool
+     */
+    protected function hasDoc(Tokens $tokens, int $index): bool
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply packages/coding-standards/src/CsFixer/Phpdoc/AbstractMissingAnnotationsFixer.php]: Will investigate

File: packages/coding-standards/src/CsFixer/Phpdoc/AbstractMissingAnnotationsFixer.php [hunk 14]
```diff
@@ -0,0 +1,317 @@
...
+     */
+    protected $phpToDocTypeTransformer;
+
+    /**
+     * @param \PhpCsFixer\WhitespacesFixerConfig $whitespacesFixerConfig
+     */
+    public function __construct(WhitespacesFixerConfig $whitespacesFixerConfig)
+    {
+        $this->whitespacesFixerConfig = $whitespacesFixerConfig;
+        $this->functionsAnalyzer = new FunctionsAnalyzer();
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply packages/coding-standards/src/CsFixer/Phpdoc/AbstractMissingAnnotationsFixer.php]: I can update the config

File: packages/coding-standards/easy-coding-standard.yml [hunk 15]
```diff
@@ -17,6 +17,10 @@ services:
     Shopsys\CodingStandards\Sniffs\ForbiddenSuperGlobalSniff: ~
     # method arguments and variables should be $camelCase
     Shopsys\CodingStandards\Sniffs\ValidVariableNameSniff: ~
+    # all methods/functions should contain all @param and @return annotations
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply packages/coding-standards/easy-coding-standard.yml]: Good point

File: composer.json [hunk 16]
```diff
@@ -41,8 +41,12 @@
...
       "Tests\\ProductFeed\\HeurekaBundle\\": "packages/product-feed-heureka/tests/",
       "Tests\\ProductFeed\\HeurekaDeliveryBundle\\": "packages/product-feed-heureka-delivery/tests/",
-      "Shopsys\\MicroserviceProductSearchExport\\": "microservices/product-search-export/src/"
-    }
+      "Shopsys\\MicroserviceProductSearchExport\\": "microservices/product-search-export/src/",
+      "Tests\\CodingStandards\\": "packages/coding-standards/tests"
+    },
+    "files": [
+      "vendor/squizlabs/php_codesniffer/autoload.php"
+    ]
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply composer.json]: That's very counter-intutive use of PHPUnit. 

It should run flawless depending on `phpunit.xml` only.

File: composer.json [hunk 17]
```diff
@@ -41,8 +41,12 @@
       "Tests\\ProductFeed\\GoogleBundle\\": "packages/product-feed-google/tests/",
       "Tests\\ProductFeed\\HeurekaBundle\\": "packages/product-feed-heureka/tests/",
       "Tests\\ProductFeed\\HeurekaDeliveryBundle\\": "packages/product-feed-heureka-delivery/tests/",
-      "Shopsys\\MicroserviceProductSearchExport\\": "microservices/product-search-export/src/"
-    }
+      "Shopsys\\MicroserviceProductSearchExport\\": "microservices/product-search-export/src/",
+      "Tests\\CodingStandards\\": "packages/coding-standards/tests"
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply composer.json]: Thanks for the fast fix PR! WOW

File: packages/coding-standards/src/CsFixer/Phpdoc/AbstractMissingAnnotationsFixer.php [hunk 18]
```diff
@@ -0,0 +1,317 @@
...
+    {
+        return false;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getName(): string
+    {
+        return ShopsysFixerNaming::createFromClass(self::class);
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply packages/coding-standards/src/CsFixer/Phpdoc/AbstractMissingAnnotationsFixer.php]: > Why do not you just return static::class?

Well, I though you have some magic related to this naming.
I personally use unique and simple `self::class` :)

I'll use the `self:class` then.

File: packages/coding-standards/src/CsFixer/Phpdoc/MissingParamAnnotationsFixer.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply packages/coding-standards/src/CsFixer/Phpdoc/MissingParamAnnotationsFixer.php]: Removed, I thought they have 2 parts that explain what the change really is.

File: packages/coding-standards/src/CsFixer/Phpdoc/OrderedParamAnnotationsFixer.php [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+
+        foreach ($lines as $key => $line) {
+            $paramName = $this->getParamNameFromLine($line);
+            if ($paramName && isset($argumentAnalyses[$paramName])) {
+                // use argument name as key, for sorting
+                $docParamNamesToKeys[$paramName] = $key;
+            }
+        }
+
+        if (!$docParamNamesToKeys) {
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply packages/coding-standards/src/CsFixer/Phpdoc/OrderedParamAnnotationsFixer.php]: You've also suggested on some places to use `count(...) === 0`.

I've picked that, as empty can accept non-iterable/non-countable elements, like `null`, `''`, `0` etc. 

Here it's always iterable, so count is more precise choice.

File: packages/coding-standards/src/CsFixer/Phpdoc/MissingParamAnnotationsFixer.php [hunk 20]
```diff
@@ -67,7 +67,7 @@ protected function processFunctionToken(Tokens $tokens, int $index, ?Token $docT
             $newLines[] = new Line(sprintf(
                 '%s * @param %s %s%s',
                 $indent,
-                $type,
+                $type ?: 'mixed',
```
Comment (TomasVotruba) [inline packages/coding-standards/src/CsFixer/Phpdoc/MissingParamAnnotationsFixer.php]: This fixes empty type to `mixed`.

In @return it doesn't make sense (could be void), so there is no type.

File: composer.json (hunk 16 above)
[the remark this answers is not in the document]
Comment (vitek-rostislav) [reply composer.json]: I see your point, however, you need to run tests of one subpackage of the monorepo only so you need to use its `phpunit.xml`. Let us discuss it :slightly_smiling_face:

File: packages/coding-standards/src/Helper/PhpToDocTypeTransformer.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (vitek-rostislav) [reply packages/coding-standards/src/Helper/PhpToDocTypeTransformer.php]: I am confused now too, I was missing the constructor annotation but here I see no arguments...

**Comment by TomasVotruba** [comment]
Applied in https://github.com/shopsys/shopsys/pull/573

**Pull request merged**